docs/internal/expressions.md

```diff
@@ -0,0 +1,597 @@
+This allows developers to modify the semantics of programs at a deeper level and makes it easy to develop DSLs on top of Verona.
+
+Verona operands, however, are similar to most languages.
+Basically: literals, variables, call arguments, return values.
+Counter-intuitively, there is no exception unwinding in Verona, exceptions thrown are just non-local return values (see below).
+
+## Literals
+
+Literals encode values of different types.
+Verona has literal of Boolean, numeric, character and array types.
+
+### Boolean
+
+Boolean literals are `true` and `false`.
+The Boolean type is `Bool` and is represented by the least-significant bit in its storage, all other bits will be ignored.
+The number of bytes in a Boolean will depend on the platform, but it will probably be at least one byte and at most one word.
+Arrays of Booleans and packing hasn't been defined yet.
+
+There is no coercion between numeric values and Boolean values, so one must initialise Boolean variables with their own literals, or as the result of Boolean expressions.
+For example:
+
+```ts
+// Direct assignment
+let a : Bool & imm = true;
+
+// This won't compile
+let b : Bool & imm = 42; // ERROR
+
+// Indirect assignment
+let c : U64 & imm = 42;
+let d : Bool & imm = (c == 0);
+
+// Using conversion from numeric types
+let e : Bool & imm = c.toBool();
+```
+
+### Numeric
+
+Both integer and floating point literals are type-inferenced by the type of the expression they're expected to match against.
+For example:
+
+```ts
+// This literal is 32-bits
+let a : I32 & mut = 42;
+
+// This one is 64-bits
+let b : F64 & mut = 3.1415;
+
+// This creates a temporary with the type that the `function` was declared with.
+function(42)
+```
+
+Numeric literals can be represented in binary (0b101010), decimal (42) or hexadecimal (0x2A) and will be treated in the same way.
+
+### Character literals
+
+Character literals are enclosed with single quotes (`'`).
+
+With the introduction of Unicode standards, characters are no longer single-bytes.
+However, the size of a character can be different depending on which Unicode standard you use (ex. `UTF-8` vs. `UTF-16`).
+But each Unicode standard has the appropriate byte-sized _escape sequence_ to represent a character with multiple bytes, so literals can still be a string of bytes.
+
+It's still under discussion if we want character literals to natively interact with numeric types.
+For example:
+
+```ts
+// `A` can be seen as 0x41, which is often OK
+let a : U8 & imm = `A`;
+
+// But `ABCD` is now a bit pattern (0x41424344)
+let b : U64 & imm = `ABCD`; // little-endian? big-endian?
+```
+
+String literals are enclosed with double quotes (`"`).
+
+String types are still in discussion, but literals can be used for their construction.
+
+For example:
+```ts
+// A plain ASCii string, initialised with a string literal
+let str : CString & imm = "Hello World!\n";
```

Comment:
This is something I discussed with Sylvan this week and have changed the example on strings. The rest is a reply to your comment on the previous iteration.

The idea of an interface that defines what a string is can be interesting for modelling abstract interfaces, but in itself it's a bit void of context. Operations on different types of strings can be wildly different and not mean the same thing (for example `size` vs `length`, or `at`, `substr` and `replace`).

More concretely, interfaces on the _actions on strings_ (ex. `Iterable`, `HumanReadable` or `RandomAccess`) are more meaningful because they tell you what you need to provide in terms of functionality, not abstract relationships.

But the real driver behind this thinking, in my view, is simplicity of implementation. Systems programming needs very simple strings to manipulate, including using as keys to maps, etc. This must be as fast as possible and trumps speed concerns for user interfaces and logs. So, a simple string class is paramount.

In that sense, a string class would be something this simple:
```
class String {
  embed _contents : Array[U8];
}
```

and other strings can make use of that in their own representations:
```
class UTF8 {
  embed _contents : String;
  ...
}
```

I haven't had such an extensive experience in handling unicode strings across different languages, so I can't tell you how much of a burden this would be. But I have had enough experience working with simple strings (`char *`) to know that an array of bytes is very ameanable to all sorts of optimisations for systems programming.

This is still an open question, but one that we can take offline and change the document once it's clearer.

Reply:
> But the real driver behind this thinking, in my view, is simplicity of implementation. Systems programming needs very simple strings to manipulate, including using as keys to maps, etc. This must be as fast as possible and trumps speed concerns for user interfaces and logs. So, a simple string class is paramount

Note: Verona is *not* a systems programming language, it is an infrastructure programming language.  The use case of keys is important, but so is handling text protocols, and parsing JSON / XML.  

> I haven't had such an extensive experience in handling unicode strings across different languages, so I can't tell you how much of a burden this would be. But I have had enough experience working with simple strings (char *) to know that an array of bytes is very ameanable to all sorts of optimisations for systems programming.

Even for kernel programming, it's painful.  For example, a POSIX filesystem generally has to support unicode for filenames and should return the same file for different representations of the same string.  This means that you have to implement unicode canonicalisation in the kernel, which adds some complexity.  

An array of code units is something that we can definitely use as a building block, but requiring contiguous storage for a string ends up leaving a lot of potential optimisation on the floor for mutable strings.  Every C/C++ codebase that does string handling ends up building some kind of twine-like abstraction because contiguous storage for mutable strings is painful, but in LLVM you have a load of redundant copies because it has to deal with interfaces that expect a C/C++ string and so has to flatten the Twine at the end.  If you define your string interfaces to be suitably flexible at the start, then you can elide this copy.  

As soon as you require a particular representation for strings, you are optimising for a specific usage pattern and you'd better be sure that that covers 100% of your use cases.  When it doesn't, as in C/C++, you end up with a proliferation of incompatible higher-level string classes and then people end up building (also incompatible) higher-level abstractions such as [UText from ICU](https://unicode-org.github.io/icu/userguide/strings/utext.html) that can wrap all of these.  If you provide a clean string interface from the start, people create a lot of different implementations of this interface and all of the code that operates on strings works across all of them.

Reply:
> The use case of keys is important, but so is handling text protocols, and parsing JSON / XML.

The way I see it is that an ASCii string is not the same thing as an UTF-8 string. They're fundamentally different classes of objects and we should be trying to common them up in an interface.

It's not because I can convert one to the other that they're similar, in the same way I can convert a floating point to an integer and vice versa, because there is often loss in the process.

Using a contiguous string of bytes for keys that have clear comparison rules is fundamental for performance, while trying to implement comparison on twine-like objects with variable sized grains is a nightmare.

But reading a JSON file as bytes and then matching symbols for parsing can be a nightmare. Not having a twine-like model here would also be a problem for streaming and incremental parsing.

So, my opinion is that we need:
* an class that is just a bag of bytes (whatever we call it), used for simple naming, map keys, symbols and low level stuff.
* a number of classes that implement human readable strings, which can themselves use the class above as storage and logic.
* potentially separate these implementations in compact / sparse modes.
* a set of explicit readers and writers (in classes like JSON and XML) on specific "string types" and not on an interface type.

> As soon as you require a particular representation for strings, you are optimising for a specific usage pattern and you'd better be sure that that covers 100% of your use cases.

Right, this is why my proposal is not to have a universal representation for "strings". But to have multiple string representations and conversion methods between them, so that developers can pick what suits them best for each case and only explicitly pay the cost of conversion.

Reply:
I'm going to "resolve" this conversation, as it has strayed too far from the intent of the document, not because what this document states is correct. There is another document on Strings and we can continue to discuss the issues there.